### DISCUSSION TITLE
#MVUX How to create a state as "not initialized"? How to validate user input?

### DISCUSSION THREAD

**Discussion by kucint** [unoplatform/uno#15349] [Q&A]
In MVU model I can create a state that can be bound to some UI element.

**Question 1:** : 
I want to create an "uninitialized" state, I mean, the UI element bound to the state shall have no value set (the textbox shall initially be empty).

Usually I create a state as follow: `public IState<int> SomeIntA => State.Value(this, () => 5);`
I can do this using `public IState<int> SomeIntB => State<int>.Empty(this);` which seems to initialize state with default value. (Am I right?)

How should I create a state and say that it is "uninitialized"?

**Question 2:** 
Let suppose user filled up the form, but left one field empty. this field is bound to some `IState<int>`.
When I read a state value associated to the field it gives me last correctly set value.
How can I detect that user did not provide a value?
Generally, how am I supposed to validate the user input?

**Comment by dr1rrb** [comment]
Hey @kucint

1. Actually I think that your issue is that you are using an `int`, so the value is exposed to the binding engine through an `int` property, so no matter how you initialize the state it will default to `0`. The solution is to create an `IState<int?>` :

| State | Initial value | Type of the generated Property | Value got by the binding engine |
| --- | --- | --- | --- |
| `State.Value(this, () => 5)` | `Some<int>(5)` | int | `5` |
| `State<int>.Empty(this)` | `None<int>()` | int | `0` |
| `State.Value<int?>(this, () => default(int?))` | `None<int?>()` | int? | `null` |
| `State<int?>.Empty(this)` | `None<int?>()` | int? | `null` |
| `State.Value<int?>(this, () => Option.Some(default(int?)))` | `Some<int?>(null)` | int? | `null` |

> Note: As shown by 3rd example, with recent versions of MVUX, we automatically convert `default(T?)` to `None`. If you want to explicitly initialize a `State` or a `Feed` with a `Some(null)` you have to use the last example in the table (not recommended).

2. Unfortunately neither WinUI nor Uno controls supports data-validation out of the box, you will have to do it by your own.
Regarding state you have 2 options, either do a `State.ForEachAsync(<callback>)`, either  simply do a `State.Select[Async](<valdiation>)` and then data-bind to the validation result..
About the validation itself, you can use the validation service https://platform.uno/docs/articles/external/uno.extensions/doc/Learn/Validation/ValidationOverview.html?tabs=notify-data-error-info#validation. E.g. : `public IFeed<IEnumerable<ValidationResult>> MyStateValidation => MyState.SelectAsync(myValue => _validator.ValidateAsync(myValue));`

**Reply by kucint** [reply]
Hi @dr1rrb 

1. it is so simple!
2. I see, I will give a try and let you know if further questions arise :)

Thanks for your quick answer!

**Reply by kucint** [reply]
I checked it out and I think I've got the logic of validation.
But I still do not understand one thing...
Let simplify the question and suppose I have a state `IState<int?> A = State<int?>.Empty(this);` 
The bounded EditBox will be initially empty. Great!
But now I want to read the value from EditBox. How should I do it?
Let's go trough couple of use cases. 
1. `int? v = await A;`
This gives me compilation warning:
_CS8714	The type 'int?' cannot be used as type parameter 'T' in the generic type or method 'Feed.GetAwaiter<T>(IFeed<T>)'. Nullability of type argument 'int?' doesn't match 'notnull' constraint._
**Something is wrong, but what exactly?**
The result of the call is `v: null;` as expected.
2. Now I put a numerical value in EditBox, let say `5` and call `int? v = await A;` again.
The result of the call is still `v: null;`. **Why?**


I guess I do something wrong, but what?

**Reply by dr1rrb** [reply]
1. The issue is that you cannot `await` a feed when the type is nullable.
The idea here is when you `await` a feed, to ease the standard case we convert a `None<T>` to `null`, but if your `T` can be `null` you would no longer be able to make the distinction between `None<T>` and `Some<T>(null)` as both would return `null`.
You can however do `await A.Option()` which will return you an `Option<T>`.
And if you are not interested to make the distinction between `None` and `Some` : `int? v = (await A.Option()).SomeOrDefault();`

2. You mentioned you are using an `EditBox`, I would assume it's actually a `TextBox`. Then your issue is because the type of [`TextBox.Text`](https://learn.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.textbox.text?view=winrt-22621) is `string` not `int`. As type does not match (and there is no implicit conversion between a `string` and a `int`) the binding engine is not able to push the value to the model, so the value remains `null`. You have then 2 options: 
    * Use a `State<string>` n your model
    * Use a [`NumberBox`](https://learn.microsoft.com/en-us/windows/winui/api/microsoft.ui.xaml.controls.numberbox.value?view=winui-2.8) in the view _(recommended)_
    > Not sure about the implicit conversion between the `double` used by the `NumberBox.Value` and a `int`. You might also have to change the type to `State<double>`.

In both cases, you will be able to remove the `?` as type will now match the type of the target property, the control will make sure to behave properly at init.


Just for curiosity, what are you trying to accomplish actually? You want user to enter a number using it's keyboard that you validate afterwards? (This is to add more details on https://github.com/unoplatform/uno.extensions/discussions/2168)

**Reply by kucint** [reply]
Hi @dr1rrb, thanks for great answer!
Yes indeed I want user to enter a number using keyboard that is afterwards validated.

But in fact, I try to achieve more general goal. I want to investigate Uno and its capabilities (C# Markup, MVUX, ...) and decide if this framework suits the company´s needs for the future product.  I seek for general answers rather to solve particular problem.
I really appreciate your input, it does really matter!

I will jump in your suggestions and post new questions if necessary :)

**Reply by kucint** [reply]
Hi @dr1rrb 
finally I found time to digest all info provided, and of course some new questions arose... 
Let assume that we have a **nullable** state A: `IState<int?> A`, as you proposed. 
I know already that if I want to read the value, I cannot `await A` directly, I must rather `await A.Option()`
What about writing the value? Usually, I use `SetAsync `or `UpdateAsync`. But:
- **Q1**:  How can I set the value? 
`await A.SetAsync(42);` causes **ERROR** _CS0453 The type 'int?' must be a non-nullable value type in order to use it as parameter 'T' in the generic type or method 'State.SetAsync<T>(IState<T>, T?, CancellationToken)'_
- **Q2**:  How can I update the value? 
`await A.UpdateAsync(p => 42);` causes **WARNING** _CS8714	The type 'int?' cannot be used as type parameter 'T' in the generic type or method 'State.UpdateAsync<T>(IState<T>, Func<T?, T?>, CancellationToken)'. Nullability of type argument 'int?' doesn't match 'notnull' constraint._
Note that `UpdateAsync` seems to works properly, in both cases:
  - when new value is 42, `await A.UpdateAsync(p => 42);`
  - when new value is null: `await A.UpdateAsync(p => null);`

**Reply by kucint** [reply]
I noticed that I can use `UpdateDataAsync`. The `SetDataAsync` does not exist, but this is not a problem....

**Reply by kucint** [reply]
Hello @dr1rrb
Bad news....
I performed many tests with nullable state values bound to many controls and I am lost. Something is generally wrong here or I am missing something. Have a look please.
I created very simple test app that contains:

- nullable state object: `public IState<int?> TextBoxIntQm => State<int?>.Empty(this);` that is bound to
- TextBox: `new TextBox().Text(x => x.Bind(() => vm.TextBoxIntQm).TwoWay())`
- and attempt to read the state value: `Option<int?> reportTextBoxIntQm = await TextBoxIntQm.Option(CancellationToken.None);`


This should work, right? But it doesn't ...
When I fill TextBox with a text `5` which definitely can be converted to `int?`, and try to read value from bound state, I get an exception:

_Error: Cannot save value from target back to source. BindingExpression: Path='TextBoxIntQm' DataItem='UnoBindingApp.BindableMainModel'; target element is 'Microsoft.UI.Xaml.Controls.TextBox' (Name='null'); target property is 'Text' (type 'String')._

Note that this exception does not crash the app, it is cough somewhere in uno/winui3, but is visible in VS output window.

I made extensive tests and this internal exception is thrown always when a nullable state is bound to any type of control (`NumberBox `behaves in the same way...)

So the question is: 
How am I supposed to read nullable value?
What am I doing wrong?
I am lost... 
Thanks for help!

BTW. if you want to get my test app (very simple one), I can expose it somehow.

**Reply by dr1rrb** [reply]
The issue is that the `TextBox.Text` property is of type `string`. So when you write `"5"` in your `TextBox` it is not able to push the value back to model `TextBoxIntQm` property which is of type `int?` (there is no implicit conversion from `string` to `int` -nor `int?`-), so when you try to read it is still on the default value.

>Note
You said that you are getting the exception only when you "try to read value from bound state", it's  probably because with a `TextBox` the text is pushed back to the model when the `TextBox` lost focus. If you try to focus any other element before reading the value, you will get the error earlier. (I'm assuming your are reading the value this by clicking on a `Button`.)

That's why I suggested you to use a `NumberBox` instead (or a `State<string>` if you really want to perform the parsing by your own).

**Reply by kucint** [reply]
I think, I cannot agree with you.
I created a test app, the purpose of that is to test how bindings between different state types and different controls behave.
States types under the test: `string, int, int? double, double? decimal, decimal?` (7 test cases)
Control types under the test: `TextBox`, `NumberBox `(2 tests cases)

All together, there is 2 x 7 = 14 combinations.

**Step 1**

Initially all states are initialized with `State.Empty()`.
The application looks as follow:

![image](https://github.com/unoplatform/uno/assets/54866088/e206930f-017f-490f-9c74-7599f27dc848)

In general all looks as expected.
- Nullable and Non-nullable states are set to `None`
- Non-nullable controls back-binding sets texts to `default` (0)
- Nullable controls back-binding sets texts set to ""
- you may notice one issue here:
  Non-nullable `decimal ` state back-binding failed to set value property.
  This is due to binding error which has a problem to convert `decimal` default value (0m) to `double` 0.0
  Binding error is to be seen in VS output window:
 _Error: Converter failed to convert value of type '0' to type 'Double'; BindingExpression: Path='StateNumberBoxDecimal' DataItem='UnoValidationApp.BindableMainModel'; target element is 'Microsoft.UI.Xaml.Controls.NumberBox' (Name='null'); target property is 'Value' (type 'Double')._ 
Generally `NumberBox` has a problem with `decimal` type (nullable or not) but this is another topic...

**Step 2**

Now I put in all fields a text "13" and try to read associated states:

![image](https://github.com/unoplatform/uno/assets/54866088/fdad4b59-4e79-4f92-a343-7309b62cd70c)

As you can see, all binding implicit conversions for non nullable types (`string, int, double` but not `decimal`) can convert correctly the text into appropriate state value. This is valid for both `TextBox` and `NumberBox`.

Unfortunately, this cannot be said about nullable states:
- `TextBox.Text` fails to convert `int?, double?, decimal?` but it converts correctly `int, double` (`decimal` fails as always->another topic...)
- `NumberBox.Value` fails to convert `int?, decimal?` but it converts correctly `int, double` and surprisingly `double?` (`decimal` fails as always...)

This can be very well seen when looking in reported binding errors in VS output window:

FYI: _Qm_ appendix means "Question mark" so StateTextBoxIntQm represents `IState<int?>`

_Error: Cannot save value from target back to source. BindingExpression: Path='StateTextBox**IntQm**' DataItem='UnoValidationApp.BindableMainModel'; target element is 'Microsoft.UI.Xaml.Controls.**TextBox**' (Name='null'); target property is 'Text' (type 'String').
Error: Cannot save value from target back to source. BindingExpression: Path='StateTextBox**DoubleQm**' DataItem='UnoValidationApp.BindableMainModel'; target element is 'Microsoft.UI.Xaml.Controls.**TextBox**' (Name='null'); target property is 'Text' (type 'String').
Error: Cannot save value from target back to source. BindingExpression: Path='StateTextBox**Decimal**' DataItem='UnoValidationApp.BindableMainModel'; target element is 'Microsoft.UI.Xaml.Controls.**TextBox**' (Name='null'); target property is 'Text' (type 'String').
Error: Cannot save value from target back to source. BindingExpression: Path='StateTextBox**DecimalQm**' DataItem='UnoValidationApp.BindableMainModel'; target element is 'Microsoft.UI.Xaml.Controls.**TextBox**' (Name='null'); target property is 'Text' (type 'String').
Error: Cannot save value from target back to source. BindingExpression: Path='StateNumberBox**IntQm**' DataItem='UnoValidationApp.BindableMainModel'; target element is 'Microsoft.UI.Xaml.Controls.**NumberBox**' (Name='null'); target property is 'Value' (type 'Double').
Error: Cannot save value from target back to source. BindingExpression: Path='StateNumberBox**Decimal**' DataItem='UnoValidationApp.BindableMainModel'; target element is 'Microsoft.UI.Xaml.Controls.**NumberBox**' (Name='null'); target property is 'Value' (type 'Double').
Error: Cannot save value from target back to source. BindingExpression: Path='StateNumberBox**DecimalQm**' DataItem='UnoValidationApp.BindableMainModel'; target element is 'Microsoft.UI.Xaml.Controls.**NumberBox**' (Name='null'); target property is 'Value' (type 'Double')._

So as you see, something is wrong here....
This test app let me to discover other issues as well, but let them tackle later....
Let us concentrate on the main topic first:
How am I suppose to read correctly nullable state values (`int?, decimal?`) from a controls (`NumberBox` and `TextBox`)? 
`double?` seems to work somehow...

**Reply by dr1rrb** [reply]
Ok I did a small repro app also on my side, and indeed the binding engine is way more permissive than I thought and allows implicit conversion from `string` to primitive types. Sorry for the confusion. And actually I noticed that uno's binding engine is even more permissive as it allows conversion to all types, including `int?` and `decimal?` the only restriction is that `NumberBox` does not work with a `decimal?` property.

Here are my result, with uno when entering "15" in all fields (and pressing the "Read" button):
![image](https://github.com/unoplatform/uno/assets/8635919/d7057022-384f-41ec-bc6c-0264249106e4)

While on windows we have issues with `int?`, `double?`, `decimal` and `decimal?`:
![image](https://github.com/unoplatform/uno/assets/8635919/ee5d1900-e188-4596-9ab5-72b0255dafca)
> The `NumberBox` works with `double?`, I think this is because the [`NumberBox.Value` is of type `double`](https://learn.microsoft.com/en-us/windows/winui/api/microsoft.ui.xaml.controls.numberbox.value?view=winui-2.8) so the conversion from `T` to `T?` is more supported

I was about to suggest you to use an `IValueConverter` to fix that (cf. collapsed section below), but I found a bug in uno with binding to nullable properties (https://github.com/unoplatform/uno/issues/16004). So I would recommend that you use an `IState<string>` and do the conversion directly in the model. Even if not optimal, it will be the best option until the issue is being fixed and which works on all platforms.

For instance :
``` csharp
public IState<string> StringProperty => State<string>.Empty(this);

public IFeed<decimal> MyDecimalProperty => StringProperty.Select(str => decimal.TryParse(str, out var result) ? result : default(decimal?));
```

You are then able to determine when the user didn't provided a value (or provided an invalid value) by doing something like this:
```csharp
if ((await MyDecimalProperty.Option(ct)).IsSome(out var value))
{
	// Value is valid
}
else
{
	// Raise validation error
}
```

Note: If you are about the have a form with multiple fields and perform validation, a good solution would be to have a record with all you fields and perform conversion and validation on the complete record. It would reduce the number of feeds involved. An example could be found here https://github.com/unoplatform/Uno.Samples/blob/master/reference/Commerce/Commerce/Commerce/Presentation/LoginViewModel.cs

BTW regarding the history of this discussion, about the nullable warning you are getting with `Feed<int?>.UpdateAsync` or when awaiting a `Feed<int?>`, it has raise some internal discussion here https://github.com/unoplatform/uno.extensions/discussions/2170, and will push soon an updated version of reactive that will allow such usages to remove friction on common scenarios like this. I've added the new extension methods in my sample so you can tests them.

We will also rename the `Feed<T>.Option()` to match the semantic "Value = `T`", "Data = `Option<T>`" and "Message = `Message<T>`" (each level gives you a deeper control of information passing through the feed).

<details>
  <summary>Solution using an `IValueConverter` blocked by https://github.com/unoplatform/uno/issues/16004</summary>

```csharp
public class GenericConvert : IValueConverter
{
	/// <inheritdoc />
	public object? Convert(object? value, Type targetType, object? parameter, string language)
		=> value is null ? null : System.Convert.ChangeType(value, targetType, CultureInfo.GetCultureInfo(language));

	/// <inheritdoc />
	public object? ConvertBack(object? value, Type targetType, object? parameter, string language)
		=> parameter switch
		{
			_ when value is null or "" => null,
			"int" => System.Convert.ToInt32(value, CultureInfo.GetCultureInfo(language)),
			"double" => System.Convert.ToDouble(value, CultureInfo.GetCultureInfo(language)),
			"decimal" => System.Convert.ToDecimal(value, CultureInfo.GetCultureInfo(language)),
			_ => value
		};
}
```
> Note: Unfortunately, windows does not provide a valid `targetType` in the `ConvertBack`, so we have to explicit it in `commandParameter`.

then it's usage:
```xaml
<Page.Resources>
	<local:SystemConvert x:Key="SystemConvert" />
	<local:GenericConvert x:Key="GenericConvert" />
</Page.Resources>

../..

<TextBox Header="int?" Text="{Binding NullableIntegerProperty, Mode=TwoWay, Converter={StaticResource GenericConvert}, ConverterParameter=int}"  />
```

</details>

**Comment by kucint** [comment]
Great! Now I see that we both get the same results, so finally we are on the same page :) 
**Thanks for your effort!**

_FYI: Project I am currently working on will not be RTM (ready to manufacture) any soon. So I do not seek for a "quick" solutions aka work-arounds. I seek rather for a long term well-structured solutions. I am ready to wait for bug-fixes even couple of months, if you assure me, they will be delivered in some predictable future._

**Question 1**: In you latest post, you attached two screenshots of your test app. The second one was done on Windows, the origin of the first one is unknown. What is the OS of the first one? Linux? macOS?

**Question 2**: Right now, bindings behave differently on Windows and other OSs (as we may see on your screenshots). Your proposed solution here is to use `IValueConverter`, which requires (https://github.com/unoplatform/uno/issues/16004) bug fix to make it work on Windows. BTW, I see that you already provided a fix for it (thanks 👍!). I suppose this bugfix will be a part of the next minor uno release, am I right?

**Question 3**: Even if using `IValueConverter` is a good work-around to make conversion work on Windows, I would rather expect more general approach here:
What is the chance to provide a fix that assures that Uno bindings behave in the same way on all OSs? I mean the behavior of `string int int? double double?`. `decimal` is another story (see next point).
I need consistent behavior on all desktop OSs, Windows behavior shall be fixed to match other OSs. 
I do not care about mobile (but probably you do...)
Is it possible? 
Shall I create a bug/issue for it?

**Question 4**: 
Now let's tackle a topic of `decimal`. This is another issue and most probably I should create another discussion for it.  If so, let me know.
The problem with `decimal` is that conversion from/to `double` is not lossless. Losing precision is not acceptable, that is why (I suppose) there is no default conversion from/to `decimal/double`.
In my opinion, the best possible work-around here would be to bind `decimal/decimal?` state to `NumberBox.Text` property rather than to `NumberBox.Value` property. Conversion from/to `string/decimal` is lossless, so it should work (on windows it doesn't but I hope it will be fixed somehow...).
Note that this use-case is not present in my test app, neither is in you test app. I have another test app that does a test of it, and unfortunately there is an issue here:
When binding a `decimal` state to a a `NumberBox.Text`, I expect that direct conversion `string/double` happens without involving `double` in it, which unfortunately seems not to happen. What really happens (I am guessing only) is: the `NumberBox` reads a `double Value` property first and converts it on the fly to `string` property. This causes precision being lost. If my suspicion is correct, the _winui3_ `NumberBox` is responsible for this misbehavior rather than uno.
I will try to deliver some screenshots soon that will present this issue in more readable way.
Question is: can the _winui3_ `NumberBox` implementation be fixed in such a way that precision loss does not happen?

**Question 4**: 
I can initialize my states to `None` as follows (no matter state type, no matter `TextBox` or `NumberBox`):
```
IState<int> StateTextBoxInt => State<int>.Empty(this);
IState<int?> StateTextBoxIntQm => State<int?>.Empty(this);
```
Let suppose I set states to 42 now:
```
await StateTextBoxInt.SetAsync(42);
await StateTextBoxIntQm.UpdateDataAsync(p => 42);  // SetDataAsync does not exist, using UpdateDataAsync instead
```
Now I want to set both states back to `None`:
- for `int` state I can write:
`await StateTextBoxInt.SetAsync(null);`                     // will result in `None`
- How am I suppose to do it for `int?` ?
`await StateTextBoxIntQm.UpdateDataAsync(null);` // will result in `Some(null)`

Thanks for your help!

**Comment by dr1rrb** [comment]
Q1: First screenshot is using skia (on windows, but it will be the same on all OS at this point: with the "skia backend" we are rendering controls by our own directly on a canvas and it can then be run on any target: Windows, Linux, macOS, IoT, etc.).

Q2: Yes, should be few weeks ahead. We're also planning to include the changes needed to await a `Feed<int?>` with next release.

Q3: On windows (i.e. using the Windows App SDK from Microsoft, not uno's skia rendering on windows), we have no control over the binding engine. And as WinAppSDK is actually our baseline, if we want to completely align the binding engine, it would mean to reduce the number of supported cases by uno. While to be purist it would make sense, as the `IValueConverter` works on all platforms, I don't see this happening any time soon.

Actually I found multiples issues on Microsoft repo about that
* https://github.com/microsoft/microsoft-ui-xaml/issues/9435
* https://github.com/microsoft/microsoft-ui-xaml/issues/8830
* https://github.com/microsoft/microsoft-ui-xaml/issues/3721

It would probably makes more sense to wait for MS to fix that on their side (and add details into the issues).

Q4:
> What really happens (I am guessing only) is: the NumberBox reads a double Value property first and converts it on the fly to string property

Damn ... but it makes sense: They are probably using `double` everywhere internally.

> can the winui3 NumberBox implementation be fixed in such a way that precision loss does not happen?

Here again it's the control from Microsoft, you will have to report it on the WinUI repro. I found that there is already an issue for that where you can add your details:
* https://github.com/microsoft/microsoft-ui-xaml/issues/2466

Then, as `NumberBox` is implemented in uno using WinUI source code, it will be fairly simple to import changes.

But usually such feature takes some time to be implemented by the WinUI team, so I would recommend you to clone [uno's implementation (which is in C#)](https://github.com/unoplatform/uno/tree/master/src/Uno.UI/Microsoft/UI/Xaml/Controls/NumberBox) to create a `DecimalNumberBox`.

Q5:
> How am I suppose to do it for int? ?

Currently you can do : `await StateTextBoxIntQm.UpdateDataAsync(_ => Option.None<int?>());` The  `Update[Data|Message]Async` methods are accepting a delegate which accepts the current value as parameter and returns the updated value. The idea behind this is to be thread safe and ensure the ACID properties of the value hold by a state (while if you get the current value, edit it and then push it back into the state, you might lose some data due to concurrent access).

For instance, this is problematic:
```csharp
public async ValueTask Increment()
{
    var current = await MyState;
    await MyState.SetAync(current + 1);
}
```
While this is thread safe (and ensure ACID properties, no matter what's hidden behind the `State`):
```csharp
public async ValueTask Increment() => await MyState.UpdateAsync(current => current + 1);
```


The `SetAsync` is a shortcut that is available only for `ValueType` and `string` as it's pretty common to write something in a `State<string>` just for data-binding and the risk of data lose is minor compared to a complex object (and that's why there is no `SetDataAsync`, it's not a common case, we prefer to fallback to the core `UpdateDataAsync` than adding too much noise in the API).

But the good news is that I pushed a commit few hours ago to add the ability to do `await state.SetAsync(null);` on an `IState<*ValueType*>` (and also on `IState<string>`) which also goes in `None`. (I pass through all `notnull` constrained operators to validate that we also have an overload for `struct` ... hopefully it should resolve a lot of your issues.)

**Reply by kucint** [reply]
_Q1: First screenshot is using skia (on windows, but it will be the same on all OS at this point: with the "skia backend" we are rendering controls by our own directly on a canvas and it can then be run on any target: Windows, Linux, macOS, IoT, etc.)._

Oh man, you open my eyes on how Uno is designed.... I was thinking that under windows it uses winUI3, under other Oss, uno emulates winUI3 somehow. Now you tell me that I can use "skia backend" as well. How can I create such a project? In the uno _dotnet new_ "wizard" there is no such a option or do I miss something?
![image](https://github.com/unoplatform/uno/assets/54866088/8b450a57-2f5d-4b6a-b084-bc2898a7f0d2)

- Product I am working on needs to run on windows (>=10), macOS, Linux. 
  "skia backend" covers all of them, right?

- It will use typical winui3 controls but as well will use Community toolkit controls, will have complex graphics using Canvas and even will provide 3D graphics using Angle.
  - "skia backend" is from the customer perspective not distinguishable from winui3 (pixel perfect?), right?
  - and for the developer as well there is no code base differences, right?

If so, this would be perfect solution to me to not be forced to test the app under all OSs.
Could you provide me more info about is "skia backend" projects, how to create them (the app and the control library)?
Thanks!

**Reply by kucint** [reply]
_Q3: On windows (i.e. using the Windows App SDK from Microsoft, not uno's skia rendering on windows), we have no control over the binding engine. And as WinAppSDK is actually our baseline, if we want to completely align the binding engine, it would mean to reduce the number of supported cases by uno. While to be purist it would make sense, as the IValueConverter works on all platforms, I don't see this happening any time soon._

I see, you are right. OK. Let stick to this solution.
you shown to me already in a previous post how to use `IValueConverter`:
```
public IState<string> StringProperty => State<string>.Empty(this);

public IFeed<decimal> MyDecimalProperty => StringProperty.Select(str => decimal.TryParse(str, out var result) ? result : default(decimal?));
```
Where exactly `IValueConverter` is involved here? I don´t see  it...
you don´t mean `StringProperty` class, do you? (BTW: where this class is defined? I cant find it...)

**Comment by kucint** [comment]
Q5
Thanks for, info, it makes much sense.

I am glad you touch the ACID properties topic, because there is couple of things I don´t understand and neither documentation nor example apps do not cover it.

Have a look, I am aware that 
```
public async ValueTask Increment()
{
    var current = await MyState;
    await MyState.SetAync(current + 1);
}
```
is problematic and should be replaced with 
`public async ValueTask Increment() => await MyState.UpdateAsync(current => current + 1);`. It is even mentioned somewhere in uno official documentation.

This scenario is quit simple, let dive in more complex one:
 
let suppose I have states: A, B, C, D
When the user clicks a button I want to Update state D which depends on A, B, C states.
How should I realize it?
